daemon/api_prompting.go

```diff
@@ -0,0 +1,381 @@
+	}
+	return uint32(userIDInt), nil
+}
+
+type interfaceManager interface {
+	AppArmorPromptingRunning() bool
+	InterfacesRequestsManager() apparmorprompting.Manager
+}
+
+var getInterfaceManager = func(c *Command) interfaceManager {
+	return c.d.overlord.InterfaceManager()
+}
+
+type postPromptBody struct {
+	Outcome     prompting.OutcomeType  `json:"action"`
+	Lifespan    prompting.LifespanType `json:"lifespan"`
+	Duration    string                 `json:"duration,omitempty"`
+	Constraints *prompting.Constraints `json:"constraints"`
+}
+
+type addRuleContents struct {
+	Snap        string                 `json:"snap"`
+	Interface   string                 `json:"interface"`
+	Constraints *prompting.Constraints `json:"constraints"`
+	Outcome     prompting.OutcomeType  `json:"outcome"`
+	Lifespan    prompting.LifespanType `json:"lifespan"`
+	Duration    string                 `json:"duration,omitempty"`
+}
+
+type removeRulesSelector struct {
+	Snap      string `json:"snap"`
+	Interface string `json:"interface,omitempty"`
+}
+
+type patchRuleContents struct {
+	Constraints *prompting.Constraints `json:"constraints,omitempty"`
+	Outcome     prompting.OutcomeType  `json:"outcome,omitempty"`
+	Lifespan    prompting.LifespanType `json:"lifespan,omitempty"`
+	Duration    string                 `json:"duration,omitempty"`
+}
+
+type postRulesRequestBody struct {
+	Action         string               `json:"action"`
+	AddRule        *addRuleContents     `json:"rule,omitempty"`
+	RemoveSelector *removeRulesSelector `json:"selector,omitempty"`
+}
+
+type postRuleRequestBody struct {
+	Action    string             `json:"action"`
+	PatchRule *patchRuleContents `json:"rule,omitempty"`
+}
+
+func getPrompts(c *Command, r *http.Request, user *auth.UserState) Response {
+	userID, errorResp := getUserID(r)
+	if errorResp != nil {
+		return errorResp
+	}
+
+	if !getInterfaceManager(c).AppArmorPromptingRunning() {
+		return InternalError("Apparmor Prompting is not running")
+	}
+
+	prompts, err := getInterfaceManager(c).InterfacesRequestsManager().Prompts(userID)
+	if err != nil {
+		return InternalError("%v", err)
+	}
+	if len(prompts) == 0 {
+		prompts = []*requestprompts.Prompt{}
+	}
+
+	return SyncResponse(prompts)
+}
+
+func getPrompt(c *Command, r *http.Request, user *auth.UserState) Response {
+	vars := muxVars(r)
+	id := vars["id"]
+
+	userID, errorResp := getUserID(r)
+	if errorResp != nil {
+		return errorResp
+	}
+
+	promptID, err := prompting.IDFromString(id)
+	if err != nil {
+		return BadRequest("%v", err)
+	}
+
+	if !getInterfaceManager(c).AppArmorPromptingRunning() {
+		return InternalError("Apparmor Prompting is not running")
+	}
+
+	prompt, err := getInterfaceManager(c).InterfacesRequestsManager().PromptWithID(userID, promptID)
+	if errors.Is(err, apparmorprompting.ErrPromptingNotEnabled) {
+		return InternalError("%v", err)
+	} else if err != nil {
+		return NotFound("%v", err)
```

Comment:
shouldn't this discriminate between `ErrNotFound` and others? `ErrClosed` is handled as an internal error below, for instance

Reply:
In the requestprompts backend there actually isn't a distinction, since prompts are stored in different data structures for each user. If we can't find the prompt in the data structure for the requested user, we return NotFound. The rationale is that this maps to the internal structure, and doesn't leak information about prompts for other users (not that this is a big deal, since existence of a prompt with a given ID for some unspecified other user is not particularly sensitive.

Here's the relevant code:
- https://github.com/canonical/snapd/blob/7562d317a5d9f22ca013a9e1c909e22d0233e023/interfaces/prompting/requestprompts/requestprompts.go#L338-L343
- https://github.com/canonical/snapd/blob/7562d317a5d9f22ca013a9e1c909e22d0233e023/interfaces/prompting/requestprompts/requestprompts.go#L353-L360
- https://github.com/canonical/snapd/blob/7562d317a5d9f22ca013a9e1c909e22d0233e023/interfaces/prompting/requestprompts/requestprompts.go#L135-L139

Reply:
Ahh good point, I didn't catch that `ErrClosed` could be returned here. Will fix.

Reply:
In fact, `ErrPromptingNotEnabled` was never actually used, and I don't think it makes a ton of sense, since the `InterfacesRequestsManager` isn't checking whether the `InterfaceManager` has prompting enabled, so this error should be removed entirely.